tc/autotuner/autotuner-inl.h

```diff
@@ -357,24 +357,62 @@ size_t largestDim(const std::vector<const DLConstTensor*>& inputs) {
 }
 
 // Creates well-chosen parameter sizes to match the input shapes.
+template <typename MappingOptionsType>
 void setupTuningParameters(
     const std::vector<const DLConstTensor*>& inputs,
-    TuningConfiguration& configuration) {
+    TuningConfiguration& configuration,
+    const std::vector<MappingOptionsType>& baseMappings) {
   TC_CHECK_GE(inputs.size(), 1u);
   auto range = inputDivisorsAndPowers2(inputs);
   // 0 is a valid tiling annotation and signals no tiling of that dimension
   // 0 is not a valid block / grid annotation
   auto nTilesDim = largestDim(inputs) + 1;
   auto tileRange = range;
   tileRange.push_back(0);
+  auto blockRange = range;
+  auto gridRange = range;
+
+  for (const auto& baseMapping : baseMappings) {
+    tileRange = mergeVectors(
+        std::move(tileRange), baseMapping.generic.tiling.extractVector());
+    blockRange =
+        mergeVectors(std::move(blockRange), baseMapping.block.extractVector());
+    gridRange =
+        mergeVectors(std::move(gridRange), baseMapping.grid.extractVector());
+  }
+
   configuration.tilingParams.setRange(nTilesDim, tileRange);
-  configuration.blockParams.setRange(range, "b");
-  configuration.gridParams.setRange(range, "g");
+  configuration.blockParams.setRange(blockRange, "b");
+  configuration.gridParams.setRange(gridRange, "g");
   configuration.unrollFactor =
       RangeParameter(powers2(FLAGS_tuner_max_unroll_size), "unroll");
   configuration.privateDepth =
       RangeParameter({0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, "pdepth");
 }
+
+// Creates well-chosen parameter sizes to match the input shapes.
+template <>
```

Comment:
Why did you chose to have the CPU version as template specialization and GPU version as default?  I would suggest having both as specializations and leaving the default template without body (link error if somebody tries it), or even doing some `enable_if` magic to get a compilation error.

You can also factor out the tiling setup to a separate function, called from both the CPU and GPU functions.